### ISSUE TITLE
How to mutate network depth/ width in NAS

### ISSUE THREAD

**Issue by NotSure2732** [microsoft/nni#4671]
**Describe the issue**:
This is not really an issue with the software per se. Rather a question which I do not know where else to ask:
While trying to make mutations to a base ResNet for a DartsTrainer I can not figure out how to use the mutations apart from for things like padding and strides.

I am trying to vary depth of a network, in that I add a LayerChoice([x,y]) where x should be do not do a layer and y should be a layer. How would I need to define x? I tried 0, null etc... but no luck. Also I can not find information in the nni documentation.

Another issue is with the width of networks. If I use LayerChoice to vary the output size e.g. 256/512 I cannot use LayerChoice in the next layer and differ the input size 256/512 because it will throw an error if the 'wrong' layer was chosen before. How could I make sure depending on which layer was chosen the right next layer will be used?

Any help is highly appreciated!

**Environment**:
- NNI version: 2.6.1
- Training service (local|remote|pai|aml|etc): local
- Client OS: MacOS
- Server OS (for remote mode only):
- Python version: 3.7
- PyTorch/TensorFlow version: 1.11.0
- Is conda/virtualenv/venv used?:
- Is running in Docker?:

**Configuration**:
 - Experiment config (remember to remove secrets!):
 - Search space:

**Log message**:
 - nnimanager.log:
 - dispatcher.log:
 - nnictl stdout and stderr:
 

**How to reproduce it?**:

**Comment by ultmaster** [comment]
For the first question, please use Repeat.

For the second question, please try ValueChoice.

Note that neither of these mutations are supported in DARTS, but we are already working on the support. For now, you can try multi-trial strategies on these.

**Comment by NotSure2732** [comment]
> Note that neither of these mutations are supported in DARTS, but we are already working on the support. For now, you can try multi-trial strategies on these.

Any update on this?

**Comment by ultmaster** [comment]
Support of ValueChoice is on master branch now. Support of Repeat is still a pending work item.

**Comment by NotSure2732** [comment]
Follow-up question:
Running DARTS on the following model (search space):

from nni.retiarii.serializer import model_wrapper
import torch.nn.functional as F
import nni.retiarii.nn.pytorch as nn
      
class Block1(nn.Module):
    def __init__(self, layer_size):
        super().__init__()
        self.conv1 = nn.Conv2d(3, layer_size, 3, stride=1,padding=1)
        self.conv2 = nn.Conv2d(layer_size, layer_size*2, 3, stride=1, padding=1)
        
    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.relu(self.conv2(x))
        return x
    
class Block2(nn.Module):
    def __init__(self, layer_size):
        super().__init__()
        self.conv1 = nn.Conv2d(3, layer_size*2, 3, stride=1,padding=1)
        
    def forward(self, x):
        x = F.relu(self.conv1(x))
        return x
    
class Block3(nn.Module):
    def __init__(self, layer_size):
        super().__init__()
        self.conv1 = nn.Conv2d(layer_size*2, layer_size*4, 3, stride=1,padding=1)
        self.conv2 = nn.Conv2d(layer_size*4, layer_size*8, 3, stride=1, padding=1)
        
    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.relu(self.conv2(x))
        return x
    
    
class Block4(nn.Module):
    def __init__(self, layer_size):
        super().__init__()
        self.conv1 = nn.Conv2d(layer_size*2, layer_size*8, 3, stride=1,padding=1)
        
    def forward(self, x):
        x = F.relu(self.conv1(x))
        return x
    
    
@model_wrapper
class Net(nn.Module):
    def __init__(self):
        super().__init__()
        rand_var = nn.ValueChoice([32,64])
        self.conv1 = nn.LayerChoice([Block1(rand_var),Block2(rand_var)])
        self.conv2 = nn.LayerChoice([Block3(rand_var),Block4(rand_var)])
        self.pool = nn.MaxPool2d(2, 2)
        self.conv3 = nn.Conv2d(rand_var*8,rand_var*16 , 3, stride=1, padding=1)
        self.fc1 = nn.Linear(rand_var*16*8*8, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)
        
    def forward(self, x):

        x = self.conv1(x)
        x = self.pool(x)
        x = self.conv2(x)
        x = self.pool(x)
        x = F.relu(self.conv3(x))  
        x = x.reshape(x.shape[0],-1)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

model = Net()


resulted in the following result:

Final architecture: {'model_2': '1', 'model_3': '1'}


The model (search space) includes 3 parameters which can vary, but the final architecture only displays 2 final choices.
Is there something wrong with the way I expressed the model space or could there be a different issue?
Since the result explicitly names 'model_2' and 'model_3' as parameters, 'model_1' would be the ValueChoice parameter...

**Comment by ultmaster** [comment]
DARTS won't scan ValueChoice nested in LayerChoice. This is the expected behavior. Please also note that there is no ValueChoice support in DARTS at all for v2.7.
